### ISSUE TITLE
Filtered AHRS stops updating

### ISSUE THREAD

**Issue by rajeshAnt** [olive-robotics/olvx_playground_imu#3]
Hi, I am having issue with getting the filtered AHRS to track the rotation. I have described the problem in detail.

### System setup ### 

- ROS Humble
- cyclone DDS
- Olive firmware: 2.2.0
- Olive software patch: 1214
- Olive DCM: 2.4.0

### Description ### 
Olive IMU is integrated to a mobile robot and while in motion the filteres_ahrs topic value seems to get static after sometime and the rotation is not tracked further, however the change in orientation is evident in the filtered_imu topic.  Is this a expected  behavior of AHRS and am I missing anything in the parameter setting? I am attaching a video to demonstrate the issue.

### example behavior ### 
[Screencast from 07-04-2024 02:58:21 PM.webm](https://github.com/olive-robotics/olv_imu_playground_cpp/assets/127131010/a2ca1ce4-968f-4d90-a525-36ebbef416fd)

As you can see the filtered AHRS value becomes static after sometime but the IMU keeps track of the rotation. And when the rotation stops AHRS corrects itself to zero quaternion while the IMU data remains static at the last rotation value.

Thanks for your help in advance.
Rajesh

**Comment by cxdcxd** [comment]
Thanks for the report, i will answer asap.

**Comment by cxdcxd** [comment]
@rajeshAnt 

The calculated yaw variable is not a standard output of the AHRS, it is a calculated value by plotjuggler. 

![image](https://github.com/olive-robotics/olv_imu_playground_cpp/assets/5897501/5dbff585-2933-4806-a08b-a33fc01ddea2)

The standard message has only quaternions. Can you print the quaternion values and check the data?

also if it is not changing please take a screenshot from logs in GUI and also from your ros2 parameters. the rig_reconfigure parameters.

**Comment by cxdcxd** [comment]
I checked the version again and couldn't find any problem. The latest version is 1215 on the website which we tuned the parameter values, other than that there is no update.

**Comment by rajeshAnt** [comment]
Hi, sorry for the late reply.

I have updated the patch to the newer version but the bug still persists. I did some further debugging and it seems to repeat whenever I subscribe to the topic for the first time, then the behavior stops and works well after sometime. The behavior can be replicated in the following way.
- restart the DCM service
- subscribe to the AHRS topic
- rotate the IMU with an approximately constant angular velocity.

Further notes: The IMU is mounted properly in a stable robot platform and rotation motion is performed on a plain terrain. 
I have attached the ros2 parameters, logs from GUI and another testing video where you see the quaternion stops and resets when i stop rotating. This behavior repeated four times in the video close to every 90 degree rotation.

[Screencast from 07-08-2024 11:10:42 AM.webm](https://github.com/olive-robotics/olv_imu_playground_cpp/assets/127131010/ee1a3ebb-40c7-4e79-8b74-9b1459448910)

![Screenshot from 2024-07-08 11-13-43](https://github.com/olive-robotics/olv_imu_playground_cpp/assets/127131010/70f6ab06-6494-4ad3-8182-6c3508e86d05)
![Screenshot from 2024-07-08 10-39-09](https://github.com/olive-robotics/olv_imu_playground_cpp/assets/127131010/3fb37c96-190d-4c00-8713-59d8e1b7819b)
![Screenshot from 2024-07-08 10-39-28](https://github.com/olive-robotics/olv_imu_playground_cpp/assets/127131010/2f764843-3a83-44cf-8277-9c0e5a97d5d5)

**Comment by cxdcxd** [comment]
@rajeshAnt 

Could we consider using only RViz2 for our visualizations? I suspect that our current issues might be related to PlotJuggler. It might be worthwhile to arrange another meeting to review everything together. As of now, I haven't pinpointed any specific problems, but a collaborative session could help us ensure everything is functioning correctly.

**Comment by cxdcxd** [comment]
![YawTest2](https://github.com/user-attachments/assets/564271ef-29dd-49a3-b258-445ad5a523bf)

This is a video recording of a 10 minutes experimnet which is speed up to 5x

**Comment by cxdcxd** [comment]
@rajeshAnt we have tested a lot with the same situation as we discussed in our meeting, but unfortunately, I couldn't replicate the problem. Can you please send me a picture of the IMU mounted on your robot? I want to see how it is exactly mounted.

**Comment by rajeshAnt** [comment]
@cxdcxd  The IMU is mounted on top of the robot like in the attached picture. I repeated the testing and i could replicate the the behavior.I have attached the video of the test.
![IMG_20240715_104449](https://github.com/user-attachments/assets/a14ef796-6dcc-48a7-9f28-2375744b4834)

The video is at 1x speed.
[Screencast from 07-15-2024 10:54:16 AM.webm](https://github.com/user-attachments/assets/879dc7fc-b810-4620-9ff2-2ede6ce9d1f7)

And the parameters are tuned like this.

![Screenshot from 2024-07-15 11-04-40](https://github.com/user-attachments/assets/720f0e45-a488-4af3-8af2-0d3c01a888e6)

can you specify your test configurations.

**Comment by rajeshAnt** [comment]
Hi @cxdcxd,

After thorough investigation, I identified the issue being related to angular velocity. Using the debug portal, I observed that the behavior recurs at specific angular velocities. It appears that any constant angular velocity below 0.04 rad/sec converges to 0 rad/sec, causing the AHRS to stop tracking the rotation. In contrast, angular velocities above 0.04 rad/sec are tracked perfectly, which suggests the problem is tied to the gyroscope's accuracy.

Given this, I conclude that the issue stems from the hardware limitations of the sensor. Therefore, I will be closing this issue.

Thank you for your support.

Best regards,
@rajeshAnt
